### ISSUE TITLE
Intellisense fails erratically for project-level lookups when project uses top-level statements

### ISSUE THREAD

**Issue by PathogenDavid** [dotnet/roslyn#50591]
**Version Used**:

* Compiler version: '3.8.0-5.20604.10 (9ed4b774)'. Language version: 9.0.
* Visual Studio 2019 (16.8.4)

Also reproduced with:

* Compiler version: '3.9.0-2.20602.2 (eb0a7ff1)'. Language version: 9.0.
* Visual Studio 2019 Preview 2.0 (16.9.0p2)

**Steps to Reproduce**:

Apologies for such a fat repro. When I strip too much out the problem goes away, so I've been unable to narrow down the exact conditions that cause this issue.

1. Recursively clone the `intellisense-with-top-level-statements-issue` branch of [my test repo](https://github.com/PathogenPlayground/InfectedImGui/tree/intellisense-with-top-level-statements-issue)
2. Open the solution (`InfectedImGui.Generator.sln`) in Visual Studio
3. Open `#TypeReferences/ImVectorTypeReference.cs`
4. Scroll to the very bottom to the `ToString` implementation on line 73.
5. As below, remove the `pe` in `ElementType`, with your cursor at the end of the property name, invoke Intellisense with Ctrl+Space. IE:

```csharp
        public override string ToString()
            => $"ImVec<{ElementTy$$}>";
```

**Expected Behavior**:

Intellisense auto-completes or suggests `ElementType`:

![image](https://user-images.githubusercontent.com/278957/104979735-9f8cc200-59ca-11eb-9dc9-752bedf515c5.png)

**Actual Behavior**:

Intellisense either finds nothing or only lists things other than the `ElementType` property:

![image](https://user-images.githubusercontent.com/278957/104979938-17f38300-59cb-11eb-9878-c80724759d4c.png)

![image](https://user-images.githubusercontent.com/278957/104979975-2cd01680-59cb-11eb-9da4-54d78da8af64.png)

----------

As stated earlier, I've had issues creating a more minimal repro for this issue. I have narrowed down the following though:

* The project must be using top-level statements
* Nullable reference types must be enabled (Either at the project level or in the file.)
* The issue only affects lookups for other types and members defined within the project (IE: Lookups for types in referenced projects work just fine.)
* The issue does not seem to happen in the top-level statements file its self.
* The issue does not go away with "Show items from unimported namespaces" disabled. (My initial reaction was this feature was malfunctioning due to it finding matching types in unimported namespaces, but that does not seem to be the case.)
* The issue happens **very** sporadically. For instance, if you convert `ToString` to use a block body, it will stop happening.
  * (I'm fairly certain I've had this happen in block bodies before too, so that's likely not part of the equation.)

You can observe that this issue is caused by the top-level statements file by opening `Program.cs` and commenting out the first `#define USE_TOP_LEVEL_STATEMENTS` line and performing the repro steps again. (Things should work as expected now.)

**Comment by PathogenDavid** [comment]
I forgot to mention that I've reproduced this on a different machine and with a clean install of Visual Studio with no extensions and default settings, so I don't believe it has anything to do with my environment.

**Comment by PathogenDavid** [comment]
It seems this issue was recently fixed as I can no longer reproduce on 16.11 or 17.0p3.1 using my instructions above. Closing this out as seemingly fixed.
